Changed calculate_voronoi_cell to use the orthogonal complement as it's artificial points.

Noel-Roemmele opened the pull request:
Fixes #38052. Fixes #37086. Changed `calculate_voronoi_cell` to use the orthogonal complement as the artificial points added to the lattice. We then multiply each element of the basis by its denominator in order to get integers instead of rationals. This solution was detailed by @DaveWitteMorris in the original issue.

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [X] The title is concise and informative.
- [X] The description explains in detail what this PR is about.
- [X] I have linked a relevant issue or discussion.
- [X] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

Commit:
Changed calculate_voronoi_cell to use the orthogonal complement as it's artificial points.

Noel-Roemmele committed:
Merge branch 'sagemath:develop' into 38052VoronoiCellOrthogonalComplement

Commit:
Buggy changes to calculate_voronoi_cell.

Commit:
Compute additional vectors using upper bound on radius of voranoi cell.

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -250,35 +250,59 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+    if radius is None:
+        from sage.rings.real_double import RDF
+        tranposedRDFMatrix = (basis.transpose()).change_ring(RDF)
+        Q, R = tranposedRDFMatrix.QR()
```
```suggestion
        R = tranposedRDFMatrix.QR()[1]
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -250,35 +250,59 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        Q, R = tranposedRDFMatrix.QR()
+        radius = 0
+        for i in range(dim[0]):
+            radius += R[i][i] * R[i][i]
```
```suggestion
        radius = sum(R[i,i]**2 for i in range(dim[0]))
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -250,35 +250,59 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+    basis = basis.LLL()
+    if radius is None:
+        from sage.rings.real_double import RDF
+        tranposedRDFMatrix = (basis.transpose()).change_ring(RDF)
```
Python does not use CamelCase, so please change `transposedRDFMatrix` to `transposed_RDF_matrix`.

Noel-Roemmele replied:
I've gotten rid of the camel case stuff.

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -250,35 +250,59 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
     if dim[0] != dim[1]:
         raise ValueError("invalid matrix")
     basis = basis / 2
+    radius = radius / 4
```
If the user supplied a `radius`, then you shouldn't divide it by 4.  Perhaps you should divide by 4 in your original calculation of `radius`. (And then you need to multiply by 4 in the formula that calculates `artificial_length`.)

Noel-Roemmele replied:
I moved the division to the original calculation of radius.

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -250,35 +250,59 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        from sage.rings.rational_field import ZZ
+        additional_vectors = additional_vectors.change_ring(ZZ)
+        additional_vectors = additional_vectors.LLL()
+        artificial_length = ceil((radius / min(approx_norm(v) for v in additional_vectors)) * 1.1)
```
Here, you want to divide by a number that is smaller than the square of the length of the shortest vector in the lattice spanned by `additional_vectors`. 

Since we need a number that is *smaller* than the shortest vector, you cannot use the `approx_norm` function, because that gives a number that is *larger* than the length. Since you don't need an integer, you could just use `v.inner_product(v)`.

Since `additional_vectors` is LLL reduced, we know that the first vector is approximately the shortest (so there is no need to look at the other vectors). However, the first vector is not necessarily shortest (and not necessarily within a factor of 1.1 of being shortest). Since the default value of `delta` in the LLL algorithm is 0.99 (actually, it might be necessary to be more careful here, because other matrix types can have a different default in their LLL algorithm), we can see from [wikipedia](https://en.wikipedia.org/wiki/Lenstra%E2%80%93Lenstra%E2%80%93Lov%C3%A1sz_lattice_basis_reduction_algorithm) that the squared length of the first vector is no more than `1.352^(n-1)` times the length of the shortest vector (where `n` is the number of additional vectors). 

And `1.1` seems a bit large as an additional multiplier to make sure that the additional vectors are strictly larger than the `radius`. How about `1.01`, or even `1.001`?

Noel-Roemmele replied:
I've used the lower bound for the shortest vector we discussed together. The one that was the minimum of the diagonals of R in the QR decomposition.

Noel-Roemmele committed:
Removed Q and use square brackets to get only necessary data.

Co-authored-by: DaveWitteMorris <43105863+DaveWitteMorris@users.noreply.github.com>

Noel-Roemmele committed:
Changed multiplication into square and use tuple to get entry.

Co-authored-by: DaveWitteMorris <43105863+DaveWitteMorris@users.noreply.github.com>

Noel-Roemmele committed:
Merge branch 'sagemath:develop' into 38052VoronoiCellOrthogonalComplement

Commit:
Fixed calculation of artificial_length.

Commit:
Added tests.

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        # Convert the basis matrix to use RDF numbers for efficiency when we
+        # perform the QR decomposition.
+        tranposed_RDF_matrix = (basis.transpose()).change_ring(RDF)
+        # We then perform the QR decomposition.
```
```suggestion
        # Convert the basis matrix to use RDF numbers for efficiency when we
        # calculate the triangular matrix of the QR decomposition.
        from sage.rings.real_double import RDF
        tranposed_RDF_matrix = (basis.transpose()).change_ring(RDF)
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
-            return r + (r1 > 0)
-        artificial_length = max(approx_norm(v) for v in basis) * 2
-        additional_vectors = identity_matrix(dim[1]) * artificial_length
+        from sage.rings.rational_field import QQ
```
```suggestion
        F = basis.base_ring().quotient_field()
```
We might break some code if we assume that the lattice consists only of integer vectors.

Noel-Roemmele replied:
`quotient_field()` doesn't seem to be a function when I try to run it in sage. The tests fail when I make that change and `ZZ.quotient_field()` also doesn't work.

DaveWitteMorris replied:
Sorry, I said the wrong thing.  It's `fraction_field` not `quotient_field`.

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
-        additional_vectors = identity_matrix(dim[1]) * artificial_length
+        from sage.rings.rational_field import QQ
+        # Introduce "artificial" basis points (representing infinity).
+        additional_vectors = (QQ**dim[1]).subspace(basis).complement().basis()
```
```suggestion
        additional_vectors = (F**dim[1]).subspace(basis).complement().basis()
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        # Convert the basis matrix to use RDF numbers for efficiency when we
+        # perform the QR decomposition.
+        tranposed_RDF_matrix = (additional_vectors.transpose()).change_ring(RDF)
+        # We then perform the QR decomposition.
```
```suggestion
 ```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        # We then perform the QR decomposition.
+        R = tranposed_RDF_matrix.QR()[1]
+        # We then get a lower bound for the shortest vector by finding the
+        # minimum all the diagonal entries of our artificial points.
```
```suggestion
        # Since R is triangular, its smallest diagonal entry provides a
        # lower bound on the length of the shortest nonzero vector in the 
        # lattice spanned by the artificial points. We square it because
        # value of `radius` is a squared length.
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        R = tranposed_RDF_matrix.QR()[1]
+        # We then get a lower bound for the shortest vector by finding the
+        # minimum all the diagonal entries of our artificial points.
+        shortest_vector_lower_bound = min(abs(R[i,i]) for i in range(dim[1] - dim[0]))
```
```suggestion
        shortest_vector_lower_bound = min(R[i,i]**2 for i in range(dim[1] - dim[0]))
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        # minimum all the diagonal entries of our artificial points.
+        shortest_vector_lower_bound = min(abs(R[i,i]) for i in range(dim[1] - dim[0]))
+        # The length we need to multiply our artificial points by in order to make
+        # sure they are greater than our radius is then calculated.
```
```suggestion
        # We will multiply our artificial points by the following scalar in
        # order to make sure the squared length of the shortest
        # nonzero vector is greater than radius, even after the vectors
        # are divided by 2.
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        sage: from sage.modules.free_module_integer import IntegerLattice
+        sage: l  = [7, 0, -1, -2, -1, -2, 7, -2, 0, 0, -2, 0, 7, -2, 0, -1, -2, -1, 7, 0 , -1, -1, 0, -2, 7]
+        sage: M = matrix(5, 5, l)
+        sage: c = IntegerLattice(M).voronoi_cell(radius=35)
```
```suggestion
        sage: C = IntegerLattice(M).voronoi_cell(radius=35)
```

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        sage: c = IntegerLattice(M).voronoi_cell(radius=35)
+        sage: M.det()
+        12858
+        sage: c.volume()
```
```suggestion
        sage: C
        A 5-dimensional polyhedron in QQ^5 defined as the
        convex hull of 720 vertices
```
As you said, this doctest was too long. We can make it much faster by eliminating the calculation of volume, but it still might be too long.

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        tranposed_RDF_matrix = (basis.transpose()).change_ring(RDF)
+        # We then perform the QR decomposition.
+        R = tranposed_RDF_matrix.QR()[1]
+        # The radius is then an upper bound for the covering radius.
```
```suggestion
        # The length of the vector formed by the diagonal entries of R is an
        # upper bound for twice the covering radius, so it is an upper bound
        # on the length of the lattice vectors that need to be considered for
        # diamond cutting. However, the value of the `radius` keyword is
        # actually a squared length, so there is no square root in the
        # following formula.
```

Noel-Roemmele committed:
Changed position of import.

Co-authored-by: DaveWitteMorris <43105863+DaveWitteMorris@users.noreply.github.com>

Commit:
Fixed lower bound calculation and changed comments.

Commit:
Added fraction_field as the base ring so that we don't break code.

Commit:
Change some documentation. Removed some code duplication in tests. Changed radius in test. Fixed artificial_length
calculation by adding a square root. Removed the conversion of additional_vectors to integers.

DaveWitteMorris commented on src/sage/modules/diamond_cutting.py:
```diff
@@ -259,19 +269,87 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
+        # following formula.
+        radius = sum(R[i,i]**2 for i in range(dim[0]))
+        # We then divide by 4 as we will divide the basis by 2 later on.
+        radius = radius / 4
```
```suggestion
        radius = ceil(radius / 4)
```
The radius should be an integer (or rational number), not `RDF`.

Noel-Roemmele committed:
Merge branch 'sagemath:develop' into 38052VoronoiCellOrthogonalComplement

Noel-Roemmele committed:
Take the ceiling oh the radius so it stays an integer.

Co-authored-by: DaveWitteMorris <43105863+DaveWitteMorris@users.noreply.github.com>
